File: src/mcsphere.c
/****************************************************************************

    mcsphere: Monte-Carlo equilibration of particles in a spherical shell 
            (Thomson's problem) for homogeneous distance between vertexes

    Copyright (C) 2019  Cesar L. Pastrana

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.

****************************************************************************/




#include "mcsphere.h"




/* Compile with: gcc mcsphere.c rands.c utils.c -std=c99 -lm --fast-math -fopenmp -O3 -o mcsphere */



int main(int argc, char *argv[])
{
    
    double *rx, *ry, *rz, *E;
    const double R = 1.0;           /* Radious of the Sphere */
    
    if(argc <3) {
        printf("Not sufficient parameters. Usage mcsphere n_particles N_MC_steps\n");
        exit(1);
    }
    
    int N = atoi(argv[1]);                    /* Number of particles */
    int MCSTEPS = atoi(argv[2]);              /* Number of Monte Carlo steps */
    
    rx = (double *)malloc(N*sizeof(double*)); 
    ry = (double *)malloc(N*sizeof(double*));
    rz = (double *)malloc(N*sizeof(double*));
    E =  (double *)malloc(MCSTEPS*sizeof(double*));
    
    printf("Minimization start (N = %d, MC_STEPS = %d: ", N, MCSTEPS);
    sphere(rx, ry, rz, E, R, N,  MCSTEPS);
    
    free(rx); 
    free(ry); 
    free(rz);
    free(E);
    
    return 0;
}


void sphere(double *rx, double *ry, double *rz, double *E, double R, int N, int MC_STEPS)
{
    double *azi, *polar;
    double tx, ty, tz, tpolar, tazi;
    double tE, DE;
    double r;
    double T, dT, kBT, BETA;
    
    int tp;
    long *idum, foornd = -time(NULL);   /* Varibles for randm number generator */   
    idum = &foornd;
    
    char numid[20];    
    char coord_flenme[1000] = COORDS_FILE_NAME;

    FILE *coords_file;
    FILE *energy_file;   
    
    T = INIT_TEMP;
    dT = (INIT_TEMP-1.0)/( (double)MC_STEPS/ANNEAL_TIME );
    kBT=kB*T;
    BETA = 1.0/kBT;
    
    
    /* Initial random positioning of the particles in the surface of a hemisphere */
    azi   = (double *)malloc(N*sizeof(double*)); 
    polar = (double *)malloc(N*sizeof(double*)); 
    for( int i=0; i<N; i++)
    {
        azi[i] = 2*PI*ran1(idum);
        polar[i] = PI*ran1(idum);	
        rx[i]=R*sin(polar[i])*cos(azi[i]);
        ry[i]=R*sin(polar[i])*sin(azi[i]);
        rz[i]=R*cos(polar[i]); 
    }
    
    /* Initial configuration energy */
    tE = 0;
    #pragma omp parallel for reduction(+:tE), private(r)
    for(int i=0; i<N; i++) {
        for(int j=0; j<N; j++) {
            if(i!=j) {
                r = edist( rx[i] - rx[j], ry[i] - ry[j], rz[i] - rz[j] );
                tE += 1.0/r;
            }
        }
    }
    E[0]=tE;
    
    
    /* +++++++++++++++++++++    MAIN MONTE CARLO ROUTINE  +++++++++++++++++++++ */
    for(int n=1; n<MC_STEPS; n++)
    {
        if(n % ANNEAL_TIME==0) { 
            T-=dT; 
            kBT = kB*T; 
            BETA = 1.0/kBT;
            
            // Here you can add if the slope is nearly constant and scape in affirmative case
        }
                
        // Select a random particle and atempt a random move 
        tp = (N - 1)*ran1(idum);
        
        /* Attempted displacement */
        tpolar = polar[tp];
        tazi   = azi[tp];
        tx = rx[tp]; 
        ty = ry[tp]; 
        tz = rz[tp];
        
        polar[tp] += Drad*(ran1(idum)-0.5);
        azi[tp] += Drad*(ran1(idum)-0.5);
        rx[tp] = R*sin(polar[tp])*cos(azi[tp]);
        ry[tp] = R*sin(polar[tp])*sin(azi[tp]);
        rz[tp] = R*cos(polar[tp]);
        

        /* Calculates the energy of the new configuration */
        tE = 0;
        #pragma omp parallel for reduction(+:tE), private(r)
        for(int i=0; i<N; i++) {
            for(int j=0; j<N; j++) {
                if(i!=j) {
                    r = edist( rx[i] - rx[j], ry[i] - ry[j], rz[i] - rz[j] );
                    tE += 1.0/r;
                }
            }
        }
        
        /* Metropolis criterion */
        DE = tE - E[n-1];
        if(  ran1(idum) <= exp(-DE*BETA) )
            E[n] = tE;
        else{
            rx[tp] = tx; 
            ry[tp] = ty; 
            rz[tp] = tz;
            azi[tp] = tazi;
            polar[tp] = tpolar;
            E[n] = E[n-1];                   
    }
    
        /* Saves the coordinates of the system after MC optimization */
        /*
        if( n > MCSTEPS - 5000 ){
            strncpy(coord_flenme, COORDS_FILE_NAME, 1000);
            itoa(n, numid);
            strcat(coord_flenme, numid );      
            strcat(coord_flenme, ".dat");
            coords_file = fopen(coord_flenme, "w");  
            for(int i=0; i<N; i++)
                fprintf(coords_file, "%.10g\t%.10g\t%.10g\n", rx[i], ry[i], rz[i]);
            fclose(coords_file); 
        }*/
    }
    /* ---------------------------------------------------------------------------- */

    free(polar);
    free(azi);
    
    
    // DATA SAVING 
    /* Writes the energy of the system as a function of the MC steps */
    energy_file = fopen(MC_ENRGY_FILE_NAME, "w");
    for(int n=0; n<MC_STEPS; n+=200)
        fprintf(energy_file, "%d\t%.10g\n", n, E[n]); 
    fclose(energy_file);

    /* Writes the eq. configuration  */    
    coords_file = fopen(COORDS_FILE_NAME, "w");  
    for(int i=0; i<N; i++)
        fprintf(coords_file, "%.10g\t%.10g\t%.10g\n", rx[i], ry[i], rz[i]);

    printf("Done! Final Temp. T = %f K\n", T);
    fclose(coords_file);

    
}

File: monte_carlo/jax/mcsphere.py
import numpy as np
import jax
import jax.numpy as jnp
import tqdm

@jax.jit
def energy(r, i_idx, j_idx):
    """
        NxN energy of the particles to be executed on the GPU
    """
    locE = jnp.where(i_idx==j_idx, 0,  1./jnp.linalg.norm(r[:,None] - r, axis=2) )
    return jnp.sum(locE)/2



def mcsphere(npart, drad, mc_steps, init_temp, anneal_time):
    """
        Generates a collection of particles homogeneously 
        distributed on the surface of a spherical shell.
        
        Input:
            npart       = int. Number of particles
            drad        = float. Maximum angular displacement of the Monte Carlo moves.
            mc_steps    = int. Number of Monte carlo steps to executed
            init_temp   = float. Initial temperature
            annel_time  = int. Annealing time (steps between temperature changes)
        
        Output:
            r = np.array [npart,3]. Cartesian coordinates of the    
                vertices defining the particles on the sphercal shell

    """
    kB = 1.38064852e-23
    T = init_temp;
    dT = (T-1.0)/(mc_steps/anneal_time)
    BETA = 1.0/(kB*T)

    r = np.zeros((npart, 3), dtype=np.float32)
    tr = np.zeros(3, dtype=np.float32)
    azi = np.zeros(npart, dtype=np.float32)
    polar = np.zeros(npart, dtype=np.float32)

    # We need these for the calculation of the energy (pair-pair)
    i_idx = jnp.arange(npart)[:,None]
    j_idx = jnp.arange(npart)[:]
    jax.vmap(energy)
             
    # Initial random positioning of the particles in the surface of the sphere 
    for i in range(0, npart):
        azi[i] = 2*np.pi*np.random.rand()
        polar[i] = np.pi*np.random.rand()
        r[i,0] = np.sin(polar[i])*np.cos(azi[i])
        r[i,1] = np.sin(polar[i])*np.sin(azi[i])
        r[i,2] = jnp.cos(polar[i])
        
    # Initial configuration energy
    E_curr = energy(r, i_idx, j_idx)

    # MAIN LOOP
    for n in range(1, mc_steps):

        if n == anneal_time:
            T -= dT
            BETA = 1.0/(kB*T);

        # Select a random particle and atempt a random move 
        tp = np.random.randint(0, npart)
        
        # Attempted displacement
        tpolar = polar[tp]
        tazi   = azi[tp]
        tr = np.copy(r[tp,:])

        polar[tp] += drad*(np.random.rand()-0.5);
        azi[tp] += drad*(np.random.rand()-0.5);
        r[tp,0] = np.sin(polar[tp])*np.cos(azi[tp])
        r[tp,1] = np.sin(polar[tp])*np.sin(azi[tp])
        r[tp,2] = jnp.cos(polar[tp])

        # Calculates the energy of the new configuration 
        # applied Metropolis criterion 
        tE = energy(r, i_idx, j_idx)
        DE = tE - E_curr;
        if np.random.rand() <= np.exp(-DE*BETA):
            E_curr = tE;
        else:
            r[tp,:] = np.copy(tr)
            azi[tp] = tazi
            polar[tp] = tpolar
        
        # Adjust the steps
        if mc_steps//2:
            drad /= 2
        if 3*mc_steps//4:
            drad /= 2
        
    return r




if __name__ == "__main__":
    NP = 8_000                      # Number of particles
    DRAD = 2.0*np.pi/180.0          # Maximum anguar displacement
    MC_STEPS = 6_000_000            # Number of Monte-Carlo steps
    T0 = 300                        # Initial temperature
    ANNEAL_TIME = 50                # Annealing time (steps between temperature changes)
    
    r = mcsphere(NP, DRAD, MC_STEPS, T0, ANNEAL_TIME) 
    np.savetxt("init_coords.dat", r, delimiter="\t")
   
   
   
   
